package_managers/download_pkgs.bzl

```diff
@@ -66,81 +88,60 @@ docker attach $cid
 docker cp $cid:{installables}.tar {output}
 # Cleanup
 docker rm $cid
- """.format(image_tar=ctx.file.image_tar.short_path,
+ """.format(image_tar=image_tar.short_path,
             image_name=builder_image_name,
             installables=ctx.attr.name,
-            download_commands=_generate_download_commands(ctx),
-            output=ctx.outputs.pkg_tar.short_path,
+            download_commands=_generate_download_commands(ctx, packages, additional_repos),
+            output=output_tar.short_path,
             )
-    _run_download_script(ctx, ctx.outputs.pkg_tar, build_contents)
+    _run_download_script(ctx, build_contents, image_tar, output_tar, output_script)
     ctx.actions.write(
-        output = ctx.outputs.executable,
+        output = output_executable,
         content = build_contents,
     )
     return struct(
-        runfiles = ctx.runfiles(files = [ctx.file.image_tar, ctx.outputs.build_script]),
-        files = depset([ctx.outputs.executable])
+        runfiles = ctx.runfiles(files = [image_tar, output_script]),
+        files = depset([output_executable])
     )
 
-_download_pkgs = rule(
-    attrs = {
-        "image_tar": attr.label(
-            default = Label("//ubuntu:ubuntu_16_0_4_vanilla.tar"),
-            allow_files = True,
-            single_file = True,
-        ),
-        "packages": attr.string_list(
-            mandatory = True,
-        ),
-    },
-    executable = True,
-    outputs = {
-        "pkg_tar": "%{name}.tar",
-        "build_script": "%{name}.sh",
-    },
+_attrs = {
+    "image_tar": attr.label(
+        default = Label("//ubuntu:ubuntu_16_0_4_vanilla.tar"),
+        allow_files = True,
+        single_file = True,
+    ),
+    "packages": attr.string_list(
+        mandatory = True,
+    ),
+    "additional_repos": attr.string_list(),
+}
+
+_outputs = {
+    "pkg_tar": "%{name}.tar",
+    "build_script": "%{name}.sh",
+}
+
+download = struct(
```

Comment:
What does this struct do?

Reply:
This struct is to be loaded from another skylark rule. For example, in my rule, I do something like this:
load("@debian_docker//package_managers:download_pkgs.bzl", _download = "download")

Then I will have access to _download.attrs, _download.outputs and _download.implementation in my rule.

This is same pattern as what container_image rule is doing: https://github.com/bazelbuild/rules_docker/blob/master/container/image.bzl#L302
https://github.com/bazelbuild/rules_docker/blob/master/container/image.bzl#L21

Reply:
ah ok . Can you add a comment mentioning Exporting Download_pkgs rule for other bazel rules to depend on.

Reply:
Done.